src/base/cache/associative_cache.hh

```diff
@@ -0,0 +1,328 @@
+ * Copyright (c) 2018 Metempsy Technology Consulting
+ * All rights reserved
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __BASE_CACHE_ASSOCIATIVE_CACHE_HH__
+#define __BASE_CACHE_ASSOCIATIVE_CACHE_HH__
+
+#include <type_traits>
+#include <vector>
+
+#include "base/cache/cache_entry.hh"
+#include "base/intmath.hh"
+#include "base/logging.hh"
+#include "base/named.hh"
+#include "base/types.hh"
+#include "mem/cache/replacement_policies/base.hh"
+#include "mem/cache/replacement_policies/replaceable_entry.hh"
+#include "mem/cache/tags/indexing_policies/base.hh"
+
+namespace gem5
+{
+
+template <typename Entry>
+class AssociativeCache : public Named
+{
+    static_assert(std::is_base_of_v<CacheEntry, Entry>,
+                  "Entry should be derived from CacheEntry");
+
+    typedef replacement_policy::Base BaseReplacementPolicy;
+
+  protected:
+
+    /** Associativity of the cache. */
+    size_t associativity;
+
+    /** The replacement policy of the cache. */
+    BaseReplacementPolicy *replPolicy;
+
+    /** Indexing policy of the cache */
+    BaseIndexingPolicy *indexingPolicy;
+
+    /** The entries */
+    std::vector<Entry> entries;
+
+  private:
+
+    void
+    initParams(size_t _num_entries, size_t _assoc)
+    {
+        fatal_if(!isPowerOf2(_num_entries), "The number of entries of an "
+                 "AssociativeCache<> must be a power of 2");
+        fatal_if(!isPowerOf2(_assoc), "The associativity of an "
+                 "AssociativeCache<> must be a power of 2");
```

Comment:
This is not always the case. For example the L2 TLB in N1 core is a 5-way set associative 1280 entry [1]. I believe the condition here should be:

    fatal_if(_num_entries % assoc != 0)

[1]: https://en.wikichip.org/wiki/arm_holdings/microarchitectures/neoverse_n1

Reply:
Updated.

Reply:
FYI, the indexing policy is implemented with the restriction that the number of sets is a power of 2. Removing this fatal from here is desired, since the AssocCache does not care whether it is a pow of 2 or not; however, to support non-pow2 architectures the indexing policies will need to be slightly reworked (as in adding some assertions and masks here and there, and fixing the tagShift and setShift, possibly fixing the skewed assoc cache). Obviously, this is not a request to do such rework, just an FYI :)

Reply:
Thanks for reporting this @odanrc. I will create an Issue detailing this so that it doesn't get lost